Another reflection bomb

xeno-by opened the pull request:
Reflection is the gift that keeps on giving.

Writing the docs for reflection and macros produced another 17 shiny commits that clean up the public API and even touch the prettyprinters. 

The previous reflection pull request caused significant controversy, but I would like to convince you, our release managers, once again. 

More than a half of the PR is removal of useless methods that somehow sneaked into the public API. Another half of the remaining half is renaming of `XXXError` exceptions wrongfully called errors. Two more commits are mandatory for Scaladoc working correctly. Another two commits are not mandatory, but they help the docs to be more comprehensible.

And the last commit upgrades prettyprinting of reflection trees, accumulating the fixes I made when prepping examples for the documentation. If you insist I'm okay with letting this one go into 2.10.1.

review @gkossakowski @jsuereth

xeno-by committed:
renames macros.ParseError to ParseException

Because it's not a fatal error. Neither it should subclass Throwable.

xeno-by committed:
renames macros.TypeError to TypecheckException

Again, this is not a fatal error, so it should end with an Error,
and it should subclass not Throwable, but Exception.

Also moved the exception outside the cake to simplify error handling,
along the same lines of what've been done for parsing and reification
exceptions.

xeno-by committed:
renames macros.ReificationError to ReificationException

And again, this is not a fatal error, so it should end with an Error,
and it should subclass not Throwable, but Exception.

xeno-by committed:
removes EmptyTermName and EmptyTypeName

We have nme.EMPTY and tpnme.EMPTY for that.

xeno-by committed:
moves Attachments from api to macros

Because they are only available in macros.Universe, not in api.Universe,
therefore I'd argue that the confusion factor is stronger than the weirdness
of scala.reflect.api.Position extending scala.reflect.macros.Attachments.

xeno-by committed:
removes extraneous methods from api.Constants

The next round of scaladoc-driven cleanup kicks in.

xeno-by committed:
removes extraneous indirection from FlagValues

We don't really need that abstract type.

xeno-by committed:
removes extraneous methods from api.Scopes

It's been more than a year, and I never used these methods.

xeno-by committed:
removes extraneous methods from api.Mirrors

We decided to give up on providing symbol table traversal facilities
in the current incarnation of mirrors. Let's be consistent with ourselves.

xeno-by committed:
turns on documentation of scala.reflect.runtime

We definitely need to document scala.reflect.runtime.universe,
therefore adding scala.reflect.runtime to skipPackages was a mistake.

But then we need to make a bunch of internal classes private to reflect
or to scala. Not very pretty, but it works.

xeno-by committed:
introduces api.JavaMirrors

This trait carries mirror-related changes of the API that happen
when api.Universe transforms into api.JavaUniverse.

From a coding standpoint this is a mere rehashing of the code, but
from a documentation standpoint this provides additional insights
into what's going on in reflection.

xeno-by committed:
hides internal methods of api.Printers

xeno-by committed:
removes extraneous methods from api.Trees

Tree.hasSymbol is really too much to document for its merit.

xeno-by committed:
upgrades showRaw

Fixes the stuff that was irritating, when I was preparing examples
for reflection documentation.

Has zero impact at stability of scalac, because showRaw isn't used anywhere
in the compiler unless invoked explicitly.

xeno-by committed:
introduces macros.package.scala

For the sole reason of putting docs on it in a separate pull request,
which is being prepared elsewhere

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/692/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1401/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1401/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/692/

xeno-by committed:
removes extraneous methods from api.StandardDefinitions

Never got to use these guys, so let's better remove them.

xeno-by committed:
removes extraneous methods from api.StandardNames

nme.ROOT doesn't have much use in the public API (unlike nme.ROOTPKG).
tpnme.EMPTY duplicates a method inherited from the base class.

xeno-by commented:
Cleaned up some commits that I didn't have the guts to clean up yesterday night

paulp commented:
LGTM.

gkossakowski commented:
This have to be discussed with @jsuereth. He'll probably respond in a few days because he is at JavaOne.

soc commented:
Would it make sense to have a common supertype for all reflection-related exceptions? Java 7 even retroactively added `ReflectiveOperationException`, so there seems to be a use-case for it.

Maybe leave it for now, but add some `TODO`, so when Scala finally targets Java 7 the exceptions can start inheriting from `ReflectiveOperationException` instead of rolling our own stuff.

(Or maybe define a class with the same name now, and as targets below 7 are dropped, replace it with a type alias to `java.lang.ReflectiveOperationException`.)

jsuereth commented on src/reflect/scala/reflect/api/Mirrors.scala:
```diff
@@ -232,14 +209,6 @@ trait Mirrors { self: Universe =>
-     *  Otherwise, if the mirror represents a runtime instance class, a mirror representing the static
-     *  part of the same class.
-     */
-    override def companion: Option[ModuleMirror]
```
Does this show up elsewhere?   Seems a useful method, but I could be wrong.

xeno-by replied:
As mentioned in the commit message, we decided to abandon navigation between mirrors in 2.10.0.

For example, one of the idea was to have `reflectMethod(name: String)` directly on `InstanceMirror` instead of having to look up a method symbol first and then call `reflectMethod(meth: MethodSymbol)`. This is convenient, but has far-reaching consistency implications, so we decided to avoid this responsibility in the first release.

In this spirit, I believe we should drop `TemplateMirror.companion` and have people go through the regular process (acquire a `companion` from the underlying symbol, do `reflectModule` or `reflectClass` on the result).

jsuereth commented on src/reflect/scala/reflect/macros/package.scala:
```diff
@@ -0,0 +1,4 @@
+package scala.reflect
+
+package object macros {
```
This seems a bit lonely.

xeno-by replied:
Will be populated in the documentation pull request to follow. Didn't want to put any documentation in this PR before chatting with the doc guys.

jsuereth commented:
OK, so same terms as before.   This has a ton of small changes and documentation fixes that LGTM, but we really can't afford to break the IDE.   The new exceptions concern me for that reason alone.

I'll test this against the community build, although I might not be able to do so until Monday.

xeno-by commented:
@soc We have a common supertype for runtime reflection exceptions, called `scala.reflect.ScalaReflectionException`. The exceptions processed here are macro-related, so I'm unsure whether they need to follow this convention.

xeno-by commented:
@jsuereth Will test the stuff this very tonight.

xeno-by commented:
@jsuereth update. Getting an error in Scala IDE tests - both with my PR and with 2.10.0-wip. Emailed Iulian about the status of Scala IDE's trunk. Looks broken :(

xeno-by commented:
@dragos Okay, here's the link to the results my last test run: https://gist.github.com/3845653 - performed against the current head of master and a recent head of 2.10.0-wip (i.e. against Scala trunk, not against this PR).

I'm still getting askOption timeouts on an ubuntu VM that has 4G of RAM and 2 CPU cores allocated. What could be done about that?

P.S. Testing against the latest commit in trunk. There's a small hope that the troubles will mysteriously fix themselves, but in the meanwhile let's think what to do. Iulian, could you please take a look?

xeno-by commented:
Same errors using the current head of 2.10.0-wip

xeno-by committed:
removes strongly-typed Constant.value wrappers

These are surely not necessary. Thanks Vlad!

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/719/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1429/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/719/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1429/
